src/backend/catalog/pg_constraint.c

```diff
@@ -836,6 +836,52 @@ get_constraint_relation_oids(Oid constraint_oid, Oid *conrelid, Oid *confrelid)
 	ReleaseSysCache(tup);
 }
 
+/*
+ * get_constraint_relation_columns
+ *		Find the columns of the relations to which a constraint refers.
+ *		Returns the list of found columns.
+ */
+List *
+get_constraint_relation_columns(Oid constraint_oid)
+{
+	List	   *conkeys_list = NIL;
```

Comment:
Maybe you should use a Bitmapset instead of a List to avoid unnecessary loops later?

Reply:
"unnecessary loops" - where do you think can we remove any loop? I see only one place in ```CQueryMutators::GroupingListContainsPrimaryKey``` where we iterate over the list of conkeys. But I don't think that we can easily get rid of this iteration.

Reply:
You can use `bms_add_member` instead `lappend_int` in this function and then use `bms_is_member` instead `var->varattno == attnum` in function `GroupingListContainsPrimaryKey`

Reply:
Hmm, I guess that could be done, but it will introduce more gpdbwrappers for bms api (currently only gpdb::BmsAddMember exists). Are you Ok with it?

Reply:
> currently only gpdb::BmsAddMember exists

Added but not checked?

> Are you Ok with it?

May be, yes